ams/server/src/main/java/com/netease/arctic/server/table/executor/DataExpiringExecutor.java

```diff
@@ -0,0 +1,479 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(DataExpiringExecutor.class);
+
+  private final Duration interval;
+  private static final Set<Type.TypeID> FIELD_TYPES = Sets.newHashSet(
+      Type.TypeID.TIMESTAMP,
+      Type.TypeID.STRING,
+      Type.TypeID.LONG
+  );
+
+  public static final String  EXPIRE_TIMESTAMP_MS = "TIMESTAMP_MS";
+  public static final String  EXPIRE_TIMESTAMP_S = "TIMESTAMP_S";
+
+  protected DataExpiringExecutor(TableManager tableManager, int poolSize, Duration interval) {
+    super(tableManager, poolSize);
+    this.interval = interval;
+  }
+
+  @Override
+  protected long getNextExecutingTime(TableRuntime tableRuntime) {
+    return interval.toMillis();
+  }
+
+  @Override
+  protected boolean enabled(TableRuntime tableRuntime) {
+    return tableRuntime.getTableConfiguration().getExpiringDataConfig().isEnabled();
+  }
+
+  @Override
+  public void handleConfigChanged(TableRuntime tableRuntime, TableConfiguration originalConfig) {
+    scheduleIfNecessary(tableRuntime, getStartDelay());
+  }
+
+  @Override
+  protected void execute(TableRuntime tableRuntime) {
+    try {
+      ArcticTable arcticTable = loadTable(tableRuntime);
+      if (!tableRuntime.getTableConfiguration().getExpiringDataConfig().isEnabled()) {
+        return;
+      }
+
+      DataExpirationConfig expirationConfig =
+          new DataExpirationConfig(arcticTable, tableRuntime.getTableConfiguration().getExpiringDataConfig());
+      if (expirationConfig.retentionTime <= 0) {
+        return;
+      }
+      purgeTableFrom(arcticTable, expirationConfig,
+          Instant.now().atZone(getDefaultZoneId(expirationConfig.expirationField)).toInstant());
+    } catch (Throwable t) {
+      LOG.error("Unexpected purge error for table {} ", tableRuntime.getTableIdentifier(), t);
+    }
+  }
+
+  /**
+   * Purge data older than the specified UTC timestamp
+   * @param table Arctic table
+   * @param expirationConfig expiration configs
+   * @param instant timestamp/timestampz/long field type uses UTC, others will use the local time zone
+   */
+  protected static void purgeTableFrom(ArcticTable table, DataExpirationConfig expirationConfig, Instant instant) {
+    long expireTimestamp = instant.minusMillis(expirationConfig.retentionTime).toEpochMilli();
+    LOG.info("Expiring Data older than {} in table {} ", Instant.ofEpochMilli(expireTimestamp)
+            .atZone(getDefaultZoneId(expirationConfig.expirationField))
+            .toLocalDateTime(),
+        table.name());
+
+    purgeTableData(table, expirationConfig, expireTimestamp);
+  }
+
+  protected static ZoneId getDefaultZoneId(Types.NestedField expireField) {
+    Type type = expireField.type();
+    if (type.typeId() == Type.TypeID.STRING) {
+      return ZoneId.systemDefault();
+    }
+    return ZoneOffset.UTC;
+  }
+
+  private static TableEntriesScan fileEntriesScan(UnkeyedTable table, Expression partitionFilter, Snapshot snapshot) {
```

Comment:
I recommend not using `TableEntriesScan` anymore, as it has significant performance issues. Can we consider using `TableScan` directly to get all files, or try the `org.apache.iceberg.FindFiles` instead?